src/pgrepr/src/value/interval.rs

```diff
@@ -46,7 +46,10 @@ impl ToSql for Interval {
         // Our intervals are guaranteed to fit within SQL's min/max intervals,
         // so this is compression is guaranteed to be lossless. For details, see
         // `repr::scalar::datetime::compute_interval`.
-        let days = std::cmp::min(self.0.days() as i128, i32::MAX as i128);
+        let days = std::cmp::max(
+            std::cmp::min(self.0.days() as i128, i32::MAX as i128),
```

Comment:
Ah, yeah, this prior behavior was bad! Our min/max intervals are way bigger than that:
https://materialize.com/docs/sql/types/interval/

Seems like this just wants to error on overflow. Does PG limit you to i32::MIN/MAX days?

Reply:
Oh I didn't see your comment @sploiselle or the docs online. Postgres does limit you to i32::MIN/MAX days:

```
postgres=# SELECT INTERVAL '2147483648 days';
ERROR:  interval field value out of range: "2147483648 days"
LINE 1: SELECT INTERVAL '2147483648 days';
                        ^
postgres=# SELECT INTERVAL '-2147483649 days';
ERROR:  interval field value out of range: "-2147483649 days"
LINE 1: SELECT INTERVAL '-2147483649 days';
```

Reply:
Ahh I see, I think that makes sense and would definitely help fix the issue. It seems like we do have a check to try and prevent exceeding Postgres bounds here: https://github.com/MaterializeInc/materialize/blob/50c6550ec894ba2dc3833b42f9a0a719c934d9f8/src/repr/src/adt/interval.rs#L63-L76
It's just that the number is wrong. 193_273_528_233_599_999_999_000 nanoseconds does not equal  2147483647 hours 59 minutes 59.999999 seconds

Reply:
Ok, for change (2) we would still need this PR so we'd be able to encode things like `INTERVAL -2147483648 -48 hrs`. For change (1) we would not need this PR.

Reply:
I think we should still do 2 now because it's small and 1 is medium or big so may take days or weeks or possibly be scrapped if we discover some problem, so it's worth it to get a small fix in now.

Reply:
Ok, I'll include it in this PR since it's not a large change.

Reply:
> It's just that the number is wrong. 193_273_528_233_599_999_999_000 nanoseconds does not equal 2147483647 hours 59 minutes 59.999999 seconds

But it does equal 2147483647 days 2147483647 hours 59 minutes 59.999999 seconds.

Anyhow, sorry folks––I wrote this code a long time ago and had forgotten my thinking/motivation and it definitely should have gotten more inline comments to detail the behavior.

Part of this is the "fallout" from our underlying representation differing from PG, which treats days as separate from H:M:S, whereas we bundle them together:

In PG:

```
sean=# SELECT interval '1 day 2147483647 hours';
        interval
------------------------
 1 day 2147483647:00:00
(1 row)

sean=# SELECT interval '1 day 2147483648 hours';
ERROR:  interval field value out of range: "1 day 2147483648 hours"
LINE 1: SELECT interval '1 day 2147483648 hours';
```

This limitation/behavior seemed pretty arbitrary to me, so I let users express Postgres' rolled-up maximum (still obeying the month-date boundary), which turns out to be 2236962132 days 07:59:59.999999. This code does then takes the maximum number of days possible (i.e. saturating the `i32`), and then rolls over any remaining day-like time units into hours.

Given that this is working as intended, I think @jkosh44 's code is right and I was wrong to say it was wrong.

This all being said, I am a little too tired tonight to reason through any issues with the values' signed-ness, but that's worth thinking through because I obviously failed to in my initial implementation.

Reply:
I see what you're saying, you're right I didn't account for the days portion in the current bounds check. I think then that the current bounds check is _mostly_ correct. Only issue is that the min is off by 1 day and 1 hour.

Reply:
So the `days` variable is computed using the the `self.0.duration` field. I've inlined the `days()` method below.
```
    pub fn days(&self) -> i64 {
        self.dur_as_secs() / (60 * 60 * 24)
    }

    pub fn dur_as_secs(&self) -> i64 {
        (self.duration / 1_000_000_000) as i64
    }
``` 

If you take a look at line 53, The `ns` variable is also being initialized with the `self.0.duration` field minus the `days` variable. So that means that any amount of days that didn't fit into the `days` variable (32 bits), will be included in the `ns` variable (64 bits) and still be lossless.

Due to the bounds of the min/max allowable Interval, we know that `self.0.duration` will never be big/small enough to over/under-flow both `days` and `ns`.